decision_service/rlclientlib/factory_resolver.cc

```diff
@@ -0,0 +1,70 @@
+#include "factory_resolver.h"
+#include <type_traits>
+#include "constants.h"
+#include "restapi_data_transport.h"
+#include "vw_model.h"
+#include "err_constants.h"
+
+namespace reinforcement_learning { 
+  namespace m = model_management;
+  namespace u = utility;
+  // For proper static intialization 
+  // Check https://en.wikibooks.org/wiki/More_C++_Idioms/Nifty_Counter for explanation
+  static int init_guard;  // guranteed to be zero when loaded
+
+  // properly aligned memory for the factory object 
+  template <typename T>
+  using natural_align = std::aligned_storage<sizeof(T), alignof ( T )>;
+
+  static natural_align<dtfactory>::type dtfactory_buf;
+  static natural_align<modelfactory>::type modelfactory_buf;
+
+  // reference should point to the allocated memory to be initalized by placement new in factory_initializer::factory_initializer()
+  dtfactory& data_transport_factory = (dtfactory&)( dtfactory_buf );
+  modelfactory& model_factory = (modelfactory&)( modelfactory_buf );
+
+  factory_initializer::factory_initializer() {
+    if ( init_guard++ == 0 ) {
```

Comment:
>if ( init_guard++ == 0 ) { [](start = 4, length = 26)

can't we just do smart pointers?

Reply:
smart pointers do reference counted de-allocation.  the use case here is around order of initialization of static object.  I don't think it applies.

---
In reply to: [193487698](https://github.com/JohnLangford/vowpal_wabbit/pull/1495#discussion_r193487698) [](ancestors = 193487698)